### PR TITLE
Make protobuf codec compatible with how links are represented in IPLD

### PR DESCRIPTION
**Pull request by mildred**
Implement new format as discussed in issue #9

### CODE REVIEW

File: coding/pb/pbcodec.go [hunk 1]
```diff
@@ -95,8 +95,19 @@ func (c *decoder) Decode(v interface{}) error {
...
 
-	if data, hasdata := n["data"]; hasdata {
+	if attrsvalue, hasattrs := n["@attrs"]; hasattrs {
+		var ok bool
+		attrs, ok = attrsvalue.(ipld.Node)
+		if !ok {
+			return nil, errInvalidData
+		}
+	} else {
+		return nil, errInvalidData
```
Comment (jbenet) [inline coding/pb/pbcodec.go]: not sure-- the null object (no links and no data) could be completely empty

**Comment by jbenet** [comment]
LGTM other than the comment above, thanks @mildred

File: coding/pb/pbcodec.go (hunk 1 above)
Comment (mildred) [inline coding/pb/pbcodec.go]: The emptiest object (generated by the reverse function in the same package) I can think of would be:

``` js
@attrs: {
  links: [],
  data: ""
}
```

But we may want to be compatible with objects generated from other sources

**Comment by mildred** [comment]
How do we deal with the fact that converting JSON to protocol buffer can be lossy? When performing the conversion, we are only reading few keys and not he whole JSON. Some additional keys would be ignored and lost in the conversion.

We can have `pb2ld(ld2pb(json)) != json`

could that be a problem?

Should we restrict ld2pb function to refuse converting to protocol buffer when there are extra keys?

**Comment by jbenet** [comment]
> We can have `pb2ld(ld2pb(json)) != json`

Still a ways out. but probably we would not. hard to achieve, but very useful property.

**Comment by jbenet** [comment]
@mildred what case are you considering?

**Comment by mildred** [comment]
@jbenet I don't understand your two last messages, but we would have an issue if we assume that decoding a protobuf object can be in every case re-encoded to protobuf while keeping its hash.

It is possible to re-encode it identically, but mostly if the decoded ipld.Node aren't modified

**Comment by jbenet** [comment]
>  hard to achieve, but very useful property.

i mean that `pb2ld(ld2pb(json)) == json` is a hard to achieve but very useful property.

> what case are you considering?

in:

> could that be a problem?
> 
> Should we restrict ld2pb function to refuse converting to protocol buffer when there are extra keys?

is there a case you were considering specifically? not sure.

**Comment by jbenet** [comment]
does this currently preserve  `pb2ld(ld2pb(json)) == json` ?

**Comment by mildred** [comment]
There exists a json for which `pb2ld(ld2pb(json)) != json`

The reason is that the protobuf format will never be able to hold arbitrary JSON keys. I suspect this won't be a problem though, because the json converted to protprotobufobuf will always come from protobuf.

Said differently, no json computed using `json = pb2json(pb)` exists for which `pb2ld(ld2pb(json)) != json`

**Comment by jbenet** [comment]
Right, so the property then is:

```
ld2pb(pb2ld(pb)) == pb
```

**Comment by mildred** [comment]
Yes, this is true: `ld2pb(pb2ld(pb)) == pb`

**Comment by jbenet** [comment]
thanks @mildred

**Pull request merged**